features/268-next-cf/api.md

```diff
@@ -0,0 +1,627 @@
+- a watchdog program that wrap function to a http server
+- routing
+- scaling functions down to zero (like hibernater in skygear v1)
+- monitoring (with prometheus)
+
+Extra features of CF we may or may not use in the future:
+- async function invocation
+
+However, one key feature OpenFaas does not provide which we need in skygear is multi-tenancy. Once you can access to the OpenFaas Gateway, you can manage all functions managed by the gateway.
+
+So skygear user would not access the OpenFaas directly to manage their functions.
+
+### Setup
+
+The OpenFaas would be installed with helm, because configuration and maintainance is more managable.
+
+src: https://github.com/openfaas/faas-netes/tree/master/chart/openfaas
+
+Some key configuration:
+- external services would NOT be exposed
+  - skygear admin may need to connect to the gateway through port forwarding
+  - thus gateway basic auth may be optional
+- imagePullPolicy => IfNotPresent, to optimise for function reboot
+  - when the cluster grow, we may need to consider how to further optimise for function boot time
+- scaleFromZero => true, faasIdler.dryRun => false, to scale down function to zero when needed
+- faasIdler.inactivityDuration affects when functions would be scale down to zero, however we will have mainly two types of deployed function, short-lived functions and long-running micro-services, scaling strategy would be discussed in later sections
+
+#### Kubernetes CRD
+
+src: https://github.com/openfaas/faas-netes/tree/master/chart/openfaas#openfaas-operator--crd-controller
+
+CRD may be considered in the future, but as suggested the more reliable faas-netes controller would be used.
+
+## CF Controller
+
+This provides APIs for skygear users to manage functions under their own authority.
+
+CF Controller would delegate the operations by accessing OpenFaas Gateway directly.
+
+Here is the list of APIs:
+- List all deployed functions
+- Get configuration, status and metrics of a function
+- Create / deploy new function
+- Update the configration of a function
+  - This will trigger new deployment of the function
+- Remove a function
+- View log of a function
+- Set function version (explain in later section)
+- Add secrets
+
+CF Controller is multi-tenant. Skygear users must login to skycli beforehead.
+
+## CF Gateway
+
+CF Gateway receives request from Gateway, and
+- forward the request to corresponding function in OpenFaas
+- for configurations of a function would be carried out in the CF Gateway, e.g.
+  - access control by keys, roles, etc, the CF Gateway connect to auth gear or the db directly to do authz
+  - secrets injection (or use k8s secrets? can be discussed)
+
+### Routing to OpenFaas
+
+Function path must be uniquely defined for OpenFaas. CF Gateway will do the path rewrite, the format would be like `/{app_name}+{version_hash}+{function_name}`, e.g. `/demo-app+abcde+hello-world` when an app called `demo-app` register a function with path `/hellow-world`.
```

Comment:
It depends on which layer is ready to be exposed publicly. 

This part is missed in the spec yet. My thought is to put authn and authz in the function pod, so that OpenFaas api gatway should be safe to be publicly accessible.